### PR TITLE
Add custom_objects in the function layer_test

### PR DESCRIPTION
**Pull request by MokkeMeguru**
Add a custom_objects argument to layer_test, to enable running layer_test for custom layers.

reference: https://github.com/tensorflow/tensorflow/issues/36272

### CODE REVIEW

**Comment by MokkeMeguru** [comment]
@googlebot I signed it!

**Comment by tomerk** [comment]
Hi Mokke,

Just realized I accidentally commented on the issue instead of this PR:
Thanks for the PR! I have two thoughts about this:

layer_test is a totally internal method, we don't want external users relying on it. This change seems explicitly designed for external users of Keras implementing their own layers and trying to use layer_test.

The layer_test utility isn't an effective test of layer behavior, it just makes sure certain things don't crash. You would be better served writing your own set of unit tests that actually verify layer behavior (and that make sure training, evaluation, etc. work correctly).

That said, I'll add @yhliang2018 to see if she thinks that there are some sort of testing utilities we should be exposing to external users in Keras' API.

**Review by tomerk — CHANGES_REQUESTED** [review]
(See my above comment)

**Comment by gbaned** [comment]
@MokkeMeguru Can you please check tomerk's comments and keep us posted? Thanks!

**Comment by MokkeMeguru** [comment]
I waited   @yhliang2018  's comment. 
But I want to say "If We Want To Update Some Layer Is Something Wrong, We Need To Pass Your UNITTEST At Least." (refs. https://github.com/tensorflow/tensorflow/issues/18222)
So this pull request gives you (and us) comfortable development of Tensorflow by removing ridiculous requests.

File: tensorflow/python/keras/testing_utils.py [hunk 1]
```diff
@@ -93,7 +93,9 @@ def layer_test(layer_cls, kwargs=None, input_shape=None, input_dtype=None,
       string or integer values.
     adapt_data: Optional data for an 'adapt' call. If None, adapt() will not
       be tested for this layer. This is only relevant for PreprocessingLayers.
-
+    custom_objects: Optional Objects for your custom layer. If you write a 
+      your layer, you can use this variable.
```
Comment (yhliang2018) [inline tensorflow/python/keras/testing_utils.py]: nit: custom_objects: Optional dictionary mapping name strings to custom objects in the layer class. This is helpful to test a custom layer.

**Comment by yhliang2018** [comment]
Thanks for the PR! I think it's okay to add an extra arg to `layer_test` to make it work better with custom layers. But the two points mentioned by @tomerk still apply: we may not expose it as an public API, and for your custom layers, you may need to test more to make it behave as expected.

File: tensorflow/python/keras/testing_utils.py (hunk 1 above)
Comment (MokkeMeguru) [reply tensorflow/python/keras/testing_utils.py]: Thanks your help!

**Comment by MokkeMeguru** [comment]
reply? Should I tell you anything more?

**Comment by gbaned** [comment]
@MokkeMeguru Can you please address the reviewers comments and resolve conflicts? Thanks!

File: tensorflow/python/keras/testing_utils.py (hunk 1 above)
Comment (tomerk) [reply tensorflow/python/keras/testing_utils.py]: I think @yhliang2018 meant that you should update this description to instead say:

```
custom_objects: Optional dictionary mapping name strings to custom objects in the layer class. This is helpful for testing custom layers.
```

Then we can go ahead and approve this.

**Comment by MokkeMeguru** [comment]
Thanks, @tomerk  !  Update document about custom_objects

**Review by tomerk — DISMISSED** [review]

**Comment by MokkeMeguru** [comment]
I fixed import error for organizing imports.

**Review by yhliang2018 — DISMISSED** [review]
Thanks for the fix. LGTM.

**Comment by MokkeMeguru** [comment]
I don't know why this pr is not accepted by Unity Sanity with pylint error. 
Can you help me?

logs.

```
=== Sanity check step 2 of 15: do_pylint (Python 3 pylint) ===

ERROR_WHITELIST="^tensorflow/python/framework/function_test\.py.*\[E1123.*noinline ^tensorflow/python/platform/default/_gfile\.py.*\[E0301.*non-iterator ^tensorflow/python/platform/default/_googletest\.py.*\[E0102.*function\salready\sdefined ^tensorflow/python/feature_column/feature_column_test\.py.*\[E0110.*abstract-class-instantiated ^tensorflow/contrib/layers/python/layers/feature_column\.py.*\[E0110.*abstract-class-instantiated ^tensorflow/contrib/eager/python/evaluator\.py.*\[E0202.*method-hidden ^tensorflow/contrib/eager/python/metrics_impl\.py.*\[E0202.*method-hidden ^tensorflow/contrib/rate/rate\.py.*\[E0202.*method-hidden ^tensorflow/python/training/tracking/tracking\.py.*\[E0202.*method-hidden ^tensorflow/python/platform/gfile\.py.*\[E0301.*non-iterator ^tensorflow/python/keras/callbacks\.py.*\[E1133.*not-an-iterable ^tensorflow/python/keras/engine/base_layer.py.*\[E0203.*access-member-before-definition ^tensorflow/python/keras/layers/recurrent\.py.*\[E0203.*access-member-before-definition ^tensorflow/python/kernel_tests/constant_op_eager_test.py.*\[E0303.*invalid-length-returned ^tensorflow/python/keras/utils/data_utils.py.*\[E1102.*not-callable ^tensorflow/python/autograph/.*_py3_test\.py.*\[E0001.*syntax-error ^tensorflow/python/keras/preprocessing/image\.py.*\[E0240.*Inconsistent method resolution "
Running pylint on 2595 files with 32 parallel jobs...


pylint took 160 s

Found a whitelisted error:
  tensorflow/python/autograph/pyct/static_analysis/activity_py3_test.py:52: [E0001(syntax-error), ] invalid syntax
Found a whitelisted error:
  tensorflow/python/autograph/impl/api_test.py:450: [E0202(method-hidden), ApiTest.test_converted_call_callable_metaclass.TestMetaclass.__call__] An attribute defined in api_test line 460 hides this method
Found a whitelisted error:
  tensorflow/python/distribute/multi_worker_test_base.py:196: [E1120(no-value-for-parameter), create_in_process_cluster] No value for argument 'reason' in unbound method call
Found a whitelisted error:
  tensorflow/python/kernel_tests/constant_op_eager_test.py:250: [E0303(invalid-length-returned), ConstantTest.testInvalidLength.BadList.__len__] __len__ does not return non-negative integer
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3293: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_invalid_type] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3293: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_invalid_type] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3300: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_one_feature_column] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3300: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_one_feature_column] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3310: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_two_feature_columns] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3310: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_two_feature_columns] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3311: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_two_feature_columns] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3311: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_two_feature_columns] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3323: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_equal_keys_different_parse_spec] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3323: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_equal_keys_different_parse_spec] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3324: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_equal_keys_different_parse_spec] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3324: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_equal_keys_different_parse_spec] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3331: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_equal_keys_equal_parse_spec] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3331: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_equal_keys_equal_parse_spec] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3332: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_equal_keys_equal_parse_spec] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3332: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_equal_keys_equal_parse_spec] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3345: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_multiple_features_dict] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3345: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_multiple_features_dict] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3346: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_multiple_features_dict] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/feature_column/feature_column_test.py:3346: [E0110(abstract-class-instantiated), MakeParseExampleSpecTest.test_multiple_features_dict] Abstract class '_TestFeatureColumn' with abstract methods instantiated
Found a whitelisted error:
  tensorflow/python/keras/saving/saved_model/json_utils.py:37: [E0202(method-hidden), Encoder.default] An attribute defined in json.encoder line 158 hides this method
Found a whitelisted error:
  tensorflow/python/keras/utils/data_utils.py:604: [E1102(not-callable), terminate_keras_multiprocessing_pools] multiprocessing.active_children is not callable
Found a whitelisted error:
  tensorflow/python/keras/utils/data_utils.py:896: [E1102(not-callable), init_pool_generator] multiprocessing.current_process is not callable
Found a whitelisted error:
  tensorflow/python/keras/utils/version_utils_test.py:120: [E0110(abstract-class-instantiated), SplitUtilsTest.test_user_provided_metaclass] Abstract class 'AbstractModel' with abstract methods instantiated

FAIL: Found 2 non-whitelisted pylint errors:
tensorflow/python/keras/testing_utils.py:201: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).

tensorflow/python/keras/testing_utils.py:257: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
```

**Comment by MokkeMeguru** [comment]
this means we should fix the code, 

```
  recovered_model = keras.models.Model.from_config(
    model_config, custom_objects)
```
to 

```
  recovered_model = keras.models.Model.from_config(
      model_config, custom_objects)
```
? (two space to four space)

**Comment by MokkeMeguru** [comment]
I updated PR for passing your lint test

**Review by tomerk — APPROVED** [review]

**Comment by MokkeMeguru** [comment]
I don't know why windows build will be crash.
I watched its log, but this error is no related to my PR.

**Comment by tomerk** [comment]
I updated the description because I *think* the curly braces in the code snippets are causing issues with the merging.

@gbaned how do we restart the merging?
Here is the code snippet I removed:

Usage:
```
testing_utils.layer_test(
    SyncBatchNormalization,
    # keras.layers.BatchNormalization,
    kwargs={
      'momentum': 0.9,
      'epsilon': 0.1,
      'gamma_regularizer': keras.regularizers.l2(0.01),
      'beta_regularizer': keras.regularizers.l2(0.01)
    },
    input_shape=(3, 4, 2),
    custom_objects={'SyncBatchNormalization': SyncBatchNormalization})
```

**Comment by MokkeMeguru** [comment]
Thanks a lot!

**Pull request merged**